### ISSUE TITLE
Playground crashes when accessing variable assigned in the previous line in an assignment expression.

### ISSUE THREAD

**Issue by dscottboggs** [crystal-lang/crystal#7091]
All three of these code snippits (and I'm assuming other similar ones) act as expected when run via `crystal eval`, but crash the playground with a similar exception to below.

## Code

```crystal
[1, 2, 3, 4] << (a = 5)
puts a
```

```crystal
def func(a)
  a
end

func(a = 5)
puts a
```

```crystal
a = (3 + (b = 2))
puts b
```

## Exception

```
Error in line 4: instantiating 'Crystal::Playground::Agent#i(Int32)'

in line 4: instantiating 'Crystal::Playground::Agent#i(Int32)'

in line 5: undefined local variable or method 'a' (did you mean 'p'?)
```

**Comment by dscottboggs** [comment]
There is another similar bug which I've just discovered. This seems to indicate that something is wrong with how the parentheses are handled, not specifically assignment expressions. Here's the new example.
```crystal
begin
  ("not a number".to_i)
rescue e : Exception
  puts e.class
end
```
This throws this exception:
```
Error in line 5: instantiating 'Crystal::Playground::Agent#i(Int32)'
in line 5: instantiating 'Crystal::Playground::Agent#i(Int32)'
in line 6: read before assignment to local variable 'e'
```
Without the parentheses, this works fine. I.E:
```crystal
begin
  "not a number".to_i
rescue e : Exception
  puts e.class
end
```
outputs "ArgumentError" as expected

**Comment by Daniel-Worrall** [comment]
Just coming across this again on playground, any sort of block will do this.
```cr
{ foo = 1 }
foo
```

**Comment by straight-shoota** [comment]
I've recently closed two duplicates of this issue and try to summarise the reasons for this behaviour.

A minimal example for the typical failing case would be this:
```cr
p a = 1
a # Error: undefined local variable or method 'a' for top-level
```
Compiling such a program with a regular compiler (like `crystal eval`) would be no issue. The playground however applies instrumentation to the code before compiling it. Essentially, each line is wrapped in a block to capture its value so it can be displayed on the right column.

You can see the actual code being compiled when running the playground in verbose mode (`-v`):
```cr
# instrumented in playground
_p.i(1) do
  p(a = 1)
end
_p.i(2) do
  a
end
```

Since `a = 1` is assigned in a block, it's scoped to that block and the local variable does not exist in top level scope.

This is only a problem when an assignment is used as a method argument (in this case, of method `p`). When the top expression of a line is an assignment, the playground takes care of this:
```cr
a = 1
a
```
```cr
# instrumented in playground
a = _p.i(1) do
  1
end
_p.i(2) do
  a
end
```
It could technically work to have the same behaviour with assignments as arguments, but it's not always correct. For example, assignments can be used as arguments to macros and then they don't mean an actual assignment to a local variable.
```cr
a = 0
record Foo, a = 1
p a # should be 0, not 1
```
It's impossible to tell from the syntactical code stage whether an assignment as call argument actually assigns to a local variable or not.

So this is hard to solve.

For the time being we must live with some limitations in the playground. An immediate improvement however would be to improve the error handling. When a "read before assignment to local variable" error occurs in the playground, it could at least suggest to make sure that any assignment is not in a call argument. With a little code analysis it could even be improved by pointing out the exact location where the instrumentation breaks the code.

**Comment by dscottboggs** [comment]
Fair enough. That makes a lot of sense. I hope there's a solution someday but I understand the limitation as you've put it. However, that doesn't explain [the latter error](https://github.com/crystal-lang/crystal/issues/7091#issuecomment-445726371) that I encountered.

**Comment by straight-shoota** [comment]
@dscottboggs That seems to be a different issue. Parenthesis should have no effect there. I've created a new issue for this: #9607
